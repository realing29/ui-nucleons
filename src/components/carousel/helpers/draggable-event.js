/**
 * Нестандартное событие перетаскивания.
 */
export default class DraggableEvent {
  /**
   * Конструктор.
   * @param {Object} props Свойства.
   * @param {import('../../helpers/point').Point} props.offset Смещение контента.
   * @param {import('../../helpers/point').Point} props.client Клиентская позиция.
   */
  constructor (props) {
    this.prevented = false;
    this.offset = props.offset;
    this.client = props.client;
  }

  /**
   * Помечает событие как отмененное.
   */
  preventDefault () {
    this.prevented = true;
  }
}
