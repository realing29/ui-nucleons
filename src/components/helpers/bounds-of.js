/**
 * Возвращает ограничивающий прямоугольник элемента.
 * @param {Element} [element] Элемент.
 * @return {DOMRect|null} Ограничивающий прямоугольник.
 */
const boundsOf = element => element && element.getBoundingClientRect
  ? element.getBoundingClientRect()
  : null;

export default boundsOf;
