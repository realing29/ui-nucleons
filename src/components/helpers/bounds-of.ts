/**
 * Возвращает ограничивающий прямоугольник элемента.
 * @param element Элемент.
 * @return Ограничивающий прямоугольник.
 */
const boundsOf = (element?: HTMLElement) => element && element.getBoundingClientRect
  ? element.getBoundingClientRect()
  : null;

export default boundsOf;
