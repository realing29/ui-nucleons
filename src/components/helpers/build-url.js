import { USER_DEFAULT_PARAMS } from '../constants';

/**
 * Подставляет корректный знак перед get-параметром.
 * @param {string} url Url.
 * @param {string} param Параметр для добавления.
 * @return {string} Get-параметр с корректным знаком.
 */
export const buildGetParameter = (url, param) => {
  const parameterSign = url.includes('?') ? '&' : '?';
  return `${parameterSign}${param}`;
};
/**
 * Добавляет get-параметры.
 * @param {string} url Url.
 * @param {Object} urlParams Пользовательские параметры для формирования URL-а.
 * @return {string} Ссылка с get-параметрами.
 */
export const addGetParameters = (url, urlParams = {}) => {
  const { sort, perPage, viewtype } = urlParams;
  // Если идем на страницу поиска, или сортировка уже присутствует в урле, то опустим сортировку
  if (sort && USER_DEFAULT_PARAMS.sort !== sort && !url.includes('/search/?q=') && !url.includes('sort=')) {
    url += buildGetParameter(url, `sort=${sort}`);
  }
  if (perPage && USER_DEFAULT_PARAMS.per_page !== perPage && !url.includes('per-page=')) {
    url += buildGetParameter(url, `per-page=${perPage}`);
  }
  if (viewtype && USER_DEFAULT_PARAMS.viewtype !== viewtype && !url.includes('viewtype=')) {
    url += buildGetParameter(url, `viewtype=${viewtype}`);
  }
  return url;
};

/**
 * Формирует URL с параметрами пользователя.
 * @param {string} url Исходный URL.
 * @param {Object} urlParams Пользовательские параметры для формирования URL-а.
 * @param {boolean} external Является ли ссылкой на внешний ресурс.
 * @param {string} anchor Якорь.
 * @return {string} Ссылка с параметрами или пустая строка.
 */
export const buildURL = ({ url = '', urlParams = {}, external = false, anchor = '' } = {}) => {
  let result = '';
  if (typeof url === 'string' && url.length) {
    let linkHref = url;
    if (!external) {
      // проверка на url вида http://valentinesday.sima-land.ru/
      if (url.match(/http[s]?:\/\/(\w*.)sima-land.ru/g)) {
        linkHref = url;
      } else if (url.indexOf('sima-land.ru') > -1) {
        linkHref = `https://www.${url.substring(url.indexOf('sima-land.ru'))}`;
      }
      linkHref = addGetParameters(linkHref, urlParams);
    }
    if (typeof anchor === 'string' && anchor.length) {
      linkHref = `${linkHref}#${anchor}`;
    }
    result = linkHref;
  }
  return result;
};
