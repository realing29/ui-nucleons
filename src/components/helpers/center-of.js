import Point from './point';

/**
 * Возвращает координаты центра прямоугольника.
 * @param {DOMRect} rect Прямоугольник.
 * @return {import('./point').Point} Центр.
 */
const centerOf = rect => Point(
  (rect.right + rect.left) / 2,
  (rect.bottom + rect.top) / 2
);

export default centerOf;
