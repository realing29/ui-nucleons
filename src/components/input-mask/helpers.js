import toArray from 'lodash/toArray';
import findLastIndex from 'lodash/findLastIndex';

/**
 * Возвращает начальное значение инпута.
 * @param {Object} params Переданные параметры.
 * @param {boolean} params.alwaysShowMask Нужно ли всегда показывать маску.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @return {string} Начальное значение инпута.
 */
export const getStartValue = ({ alwaysShowMask, value, mask, pattern, maskSign }) => {
  let startValue;

  if (value || alwaysShowMask) {
    startValue = mask
      ? prepareValue({ value, mask, maskSign, pattern })
      : (value.match(pattern) || []).join('');
  }

  return startValue || '';
};

/**
 * Преобразование значения под маску.
 * @param {Object} params Переданные параметры.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @return {string} Преобразованное значение.
 */
export const prepareValue = ({ value, mask, maskSign, pattern }) => {
  const patternMask = (mask.match(pattern) || []).join('');
  const patternValueArray = value.match(pattern) || [];
  const patternValue = patternValueArray.join('');

  // Убираем цифры маски из переданного значения инпута.
  const lengthPatternMask = patternMask.length;
  lengthPatternMask
    && patternValue.indexOf(patternMask) === 0
    && patternValueArray.splice(0, lengthPatternMask);

  // Последовательно заполняем заменяемые символы маски по возможности.
  const correctValue = mask
    ? toArray(mask).map(character => {
      let newCharacter;
      if (character === maskSign) {
        newCharacter = patternValueArray.shift();
      }
      return newCharacter || character;
    }).join('')
    : patternValue;

  return correctValue || mask;
};

/**
 * Возвращает корректную позицию курсора.
 * @param {Object} params Переданные параметры.
 * @param {string} params.value Значение инпута.
 * @param {string} params.mask Маска.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @param {number} params.position Переданное положение курсора.
 * @return {number} Позиция курсора.
 */
export const validateCursorPosition = ({ value, mask, maskSign, position }) => {
  const firstValueCaret = value.indexOf(maskSign);
  const firstMaskCaret = mask.indexOf(maskSign);
  let correctPosition = position >= 0 ? position : 0;

  if (firstValueCaret >= 0 && correctPosition >= firstValueCaret) {
    correctPosition = firstValueCaret;
  }

  if (firstMaskCaret >= 0 && correctPosition <= firstMaskCaret) {
    correctPosition = firstMaskCaret;
  }

  return correctPosition;
};

/**
 * Обрабатывает изменения инпута.
 * @param {Object} params Переданные параметры.
 * @param {HTMLElement} params.input Инпут.
 * @param {string} params.mask Маска.
 * @param {RegExp} params.pattern Паттерн допустимых значений.
 * @param {string} params.maskSign Заменяемый знак маски.
 * @param {null|number} params.lastSelection Последнее значение положения курсора.
 * @return {Object} Обновленные данные значения инпута и положения курсора.
 */
export const processChange = ({ input, mask, maskSign, pattern, lastSelection }) => {
  const { value = '', selectionStart } = input || {};
  const maskArray = toArray(mask);
  const valueCharactersArray = toArray(value);
  let position = selectionStart;

  const isAdded = value.length > mask.length;

  // Расчет позиции курсора, чтоб он всегда находился перед заменяемым знаком маски.
  if (maskArray[selectionStart] !== maskSign) {
    const deltaSelection = lastSelection - selectionStart;
    const [needRemoveBefore, needRemoveAfter] = [deltaSelection === 1, deltaSelection === 0];

    if (needRemoveBefore) {
      position = findLastIndex(maskArray, (character, index) =>
        index < selectionStart && character === maskSign);
    } else if (needRemoveAfter || isAdded) {
      position = maskArray.findIndex((character, index) =>
        index >= selectionStart && character === maskSign);
    }

    // Удаление корректного значения, если при удалении попали на пробел/скобку и т.д.
    if (position >= 0 && (needRemoveBefore || needRemoveAfter)) {
      const removeIndex = needRemoveBefore ? position : position - 1;
      valueCharactersArray[removeIndex] = maskArray[removeIndex];
    }
  }

  const correctValue = prepareValue({
    value: valueCharactersArray.join(''),
    mask,
    maskSign,
    pattern,
  }) || '';

  return {
    correctValue,
    correctSelection: validateCursorPosition({
      value: correctValue,
      mask,
      maskSign,
      position: position > 0 ? position : correctValue.length,
    }),
  };
};
