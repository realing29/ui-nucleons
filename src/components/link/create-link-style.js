import classNames from 'classnames/bind';
import styles from './link.scss';

const cx = classNames.bind(styles);
export const LINK_COLORS = ['black', 'dark-blue', 'blue', 'red', 'gray', 'white'];
export const EXTERNAL_COLORS = ['blue', 'white', 'black'];
export const LINK_UNDERLINE_TYPES = ['solid', 'dashed'];

/**
 * Создание стилей ссылки/псевдо-ссылки.
 * @param {string} [className=''] Пользовательские классы.
 * @param {boolean} [disableHoverEffect=false] Не реагировать при наведении.
 * @return {string} Строка с классами для ссылки/псевдо-ссылки.
 */
export const createLinkStyle = ({ className = '', disableHoverEffect = false }) =>
  cx(className, {
    link: true,
    'link-can-be-hovered': !disableHoverEffect,
  });

/**
 * Создание стилей для текста внутри ссылки/псевдо-ссылки.
 * @param {string} [color=''] Цвет текста.
 * @param {string} [underlineType=''] Подчеркивание.
 * @param {boolean} [external=false] Является ли ссылкой на внешний ресурс.
 * @return {string} Строка с классами для текста внутри ссылки/псевдо-ссылки.
 */
export const createLinkTextStyle = ({ color = '', underlineType = '', external = false }) => {
  color = LINK_COLORS.includes(color) ? color : 'blue';
  underlineType = LINK_UNDERLINE_TYPES.includes(underlineType) ? underlineType : false;
  return cx('link-text', `link-${color}`, {
    'link-external': external,
    'link-underlined': underlineType,
    [`link-underlined-${underlineType}`]: underlineType,
  });
};

/**
 * Создание стилей для иконки внешного ресурса.
 * @param {string} [color] Цвет.
 * @return {string} Строка с классами для иконки внешнего ресурса.
 */
export const createExternalStyle = color => {
  const iconColor = color && EXTERNAL_COLORS.includes(color) ? color : 'blue';
  return cx('icon-external', `external-${iconColor}`);
};
