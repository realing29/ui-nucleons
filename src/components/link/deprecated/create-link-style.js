import classNames from 'classnames/bind';
import styles from './link.scss';

const cx = classNames.bind(styles);
export const LINK_COLORS = ['black', 'dark-blue', 'blue', 'red', 'gray', 'white', 'light-gray', 'dark-gray'];
export const LINK_UNDERLINE_TYPES = ['solid', 'dashed'];

/**
 * Создание стилей ссылки/псевдо-ссылки.
 * @param {Object} params Входные параметры.
 * @param {string} [params.className=''] Пользовательские классы.
 * @param {string} [params.color=''] Цвет текста.
 * @param {string} [params.underlineType=''] Подчеркивание.
 * @param {boolean} [params.disableHoverEffect=false] Не реагировать при наведении.
 * @return {string} Строка с классами для ссылки/псевдо-ссылки.
 */
export const createLinkStyle = ({ className = '', color = '', underlineType = '', disableHoverEffect = false }) => {
  const allowedColor = LINK_COLORS.includes(color) ? color : 'blue';
  const allowedUnderlineType = LINK_UNDERLINE_TYPES.includes(underlineType) ? underlineType : false;

  return cx(className, 'link', `link-${allowedColor}`, {
    'link-can-be-hovered': !disableHoverEffect,
    'link-underlined': allowedUnderlineType,
    [`link-underlined-${allowedUnderlineType}`]: allowedUnderlineType,
  });
};
