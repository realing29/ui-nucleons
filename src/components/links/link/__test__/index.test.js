import Link from '../';
import { shallow } from 'enzyme';
import React from 'react';
import * as styleHelpers from '../../create-link-style';
import * as urlHelper from '../../../helpers/build-url';

describe('<Link />', () => {
  it('calls helpers with right params and renders correctly without external', () => {
    jest.spyOn(urlHelper, 'buildURL');
    jest.spyOn(styleHelpers, 'createLinkStyle');
    jest.spyOn(styleHelpers, 'createLinkTextStyle');
    jest.spyOn(styleHelpers, 'createExternalStyle');
    const params = {
      sort: 'price',
    };
    const link = shallow(
      <Link
        className='testClass'
        url='/cart/'
        urlParams={params}
        anchor='test_anchor'
        color='white'
        underlined
        inline
        relative
        disableHoverEffect
      >
        TestLink
      </Link>
    );
    expect(styleHelpers.createLinkStyle).toHaveBeenCalledTimes(1);
    expect(styleHelpers.createLinkStyle).toHaveBeenCalledWith({
      className: 'testClass',
      disableHoverEffect: true,
      inline: true,
      relative: true,
    });
    expect(styleHelpers.createLinkTextStyle).toHaveBeenCalledTimes(1);
    expect(styleHelpers.createLinkTextStyle).toHaveBeenCalledWith({
      color: 'white',
      underlineType: 'solid',
    });
    expect(styleHelpers.createExternalStyle).toHaveBeenCalledTimes(1);
    expect(styleHelpers.createExternalStyle).toHaveBeenCalledWith('white');
    expect(urlHelper.buildURL).toHaveBeenCalledTimes(1);
    expect(urlHelper.buildURL).toHaveBeenCalledWith({
      anchor: 'test_anchor',
      external: undefined,
      url: '/cart/',
      urlParams: params,
    });
    expect(link).toMatchSnapshot();
  });
  it('calls helpers and renders correctly with external', () => {
    jest.spyOn(urlHelper, 'buildURL');
    jest.spyOn(styleHelpers, 'createLinkStyle');
    jest.spyOn(styleHelpers, 'createLinkTextStyle');
    jest.spyOn(styleHelpers, 'createExternalStyle');
    const params = {
      currentCurrency: {
        name: 'USD',
      },
      sort: 'price',
    };
    const link = shallow(
      <Link
        url='/cabinet/'
        urlParams={params}
        color='white'
        external
        disableHoverEffect
      >
        TestLink
      </Link>
    );
    expect(styleHelpers.createLinkStyle).toHaveBeenCalledTimes(2);
    expect(styleHelpers.createLinkTextStyle).toHaveBeenCalledTimes(2);
    expect(styleHelpers.createExternalStyle).toHaveBeenCalledTimes(2);
    expect(urlHelper.buildURL).toHaveBeenCalledTimes(2);
    expect(link).toMatchSnapshot();
  });
  it('calls functions onClick, onMouseEnter, onMouseLeave', () => {
    const onClickFunc = jest.fn(),
      onMouseEnterFunc = jest.fn(),
      onMouseLeaveFunc = jest.fn();
    const link = shallow(
      <Link
        onClick={onClickFunc}
        onMouseEnter={onMouseEnterFunc}
        onMouseLeave={onMouseLeaveFunc}
      >
        TestLink
      </Link>
    );
    link.simulate('click');
    expect(onClickFunc).toHaveBeenCalledTimes(1);
    link.simulate('mouseEnter');
    expect(onMouseEnterFunc).toHaveBeenCalledTimes(1);
    link.simulate('mouseLeave');
    expect(onMouseLeaveFunc).toHaveBeenCalledTimes(1);
  });
});
