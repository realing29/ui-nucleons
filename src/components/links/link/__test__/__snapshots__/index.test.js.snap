// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> calls helpers and renders correctly with external 1`] = `
<a
  className="link link-block"
  href="/cabinet/"
  target="_blank"
>
  <span
    className="link-text link-white link-external"
  >
    TestLink
  </span>
  <span
    className="icon-external external-white"
  >
    <Icon
      icon={[Function]}
      inline={true}
      size={16}
    />
  </span>
</a>
`;

exports[`<Link /> calls helpers with right params and renders correctly without external 1`] = `
<a
  className="testClass link link-relative"
  href="/cart/?sort=price#test_anchor"
  target="_self"
>
  <span
    className="link-text link-white link-underlined link-underlined-solid"
  >
    TestLink
  </span>
</a>
`;
